Bug fixes in WCSAxes.plot_coord

Cadair opened the pull request:
This fixes two bugs in `plot_coord`:

1) It more correctly uses `.spherical` rather than `.data` to get `lon` and `lat`.
2) It does the frame transform on the input coordinate frame rather than relying on the matplotlib transforms later.

The latter is a fix because the matplotlib transforms are only ever two dimensional, and it should not be assumed that the input coordinate to `plot_coord` is two dimensional, as it can be in any frame transformable to the native frame of the plot.

I can't really think of a way to test this with just the astropy frames, if anyone has any suggestions of a transformation which would work in 3D but not 2D in astropy frames that would be great.

ping @ayshih

pllim commented:
Is this addressing any existing Astropy issue(s)? Also please add change log and tests. Thanks!

bsipocz commented on CHANGES.rst:
```diff
@@ -371,6 +371,8 @@ astropy.utils
 astropy.visualization
 ^^^^^^^^^^^^^^^^^^^^^
 
+- Fix incorrect transformation behavior in ``WCSAxes.plot_coord`` and correctly handle when input coordinates are not already in spherical representations. [#8927]
```
wrap the line please (not critical, but would be nice :) )

astrofrog commented:
Possible regression test (fails pretty badly with master):

```
from astropy import units as u
from astropy.wcs import WCS
from astropy.coordinates import SkyCoord
from astropy.utils.data import get_pkg_data_filename

import matplotlib.pyplot as plt

filename = get_pkg_data_filename('galactic_center/gc_msx_e.fits')
wcs = WCS(filename)

coord = SkyCoord(0 * u.kpc, 0 * u.kpc, 0 * u.kpc, frame='galactocentric')

fig = plt.figure()
ax = fig.add_subplot(1, 1, 1, projection=wcs)
ax.plot_coord(coord, 'ro')
fig.savefig('galcen.png')
```

The marker should be at the center of the image.

Cadair commented on astropy/visualization/wcsaxes/core.py:
```diff
@@ -299,7 +304,7 @@ def plot_coord(self, *args, **kwargs):
                                 " as it is automatically determined by the input coordinate frame.")
 
-            transform = self.get_transform(frame0)
+            transform = self.get_transform(native_frame)
```
I did that at one point but this seemed easier to read. Is there a good reason for one or the other?

Cadair commented on astropy/visualization/wcsaxes/core.py:
```diff
@@ -285,12 +286,16 @@ def plot_coord(self, *args, **kwargs):
+                    plot_data.append(frame0.spherical.lon.to_value(u.deg))
                 elif coord.coord_type == 'latitude':
-                    plot_data.append(frame0.data.lat.to_value(coord.coord_unit))
+                    plot_data.append(frame0.spherical.lat.to_value(u.deg))
```
I think that frames cache their different representations anyway, so this should be basically free on both counts.

Cadair commented:
@astrofrog Not sure what the procedure is for the image tests?

astrofrog commented:
The image test docs are here: http://docs.astropy.org/en/stable/development/testguide.html#image-tests-with-pytest-mpl

However, can't you just check the pixel coordinates of the resulting marker without doing a full image test? (which seems overkill here)

Cadair commented:
@astrofrog I changed the test. Can you take a look to check I am comparing it to the right thing?

Also I changed the isort config because I needed to use isort.

astrofrog commented on astropy/visualization/wcsaxes/tests/test_display_world_coordinates.py:
```diff
@@ -117,3 +123,18 @@ def test_cube_coords(self, tmpdir):
+        ax.imshow(data)
+        point, = ax.plot_coord(coord, 'ro')
+
+        np.testing.assert_allclose(point.get_xydata()[0], [0, 0], atol=1e-4)
```
I'm surprised that this would be at [0,0], it should be near the center of the image.

Cadair replied:
yeah this one confused me as well. is `get_xydata` in pixels?

astrofrog replied:
Looks like it should be 😕

astrofrog replied:
Ah no I think it might be the data before applying the ``transform``. So to some extent this test is correct but I'd still be happier if we can figure out how to get the transformed data.

astrofrog replied:
You might need ``point.get_transform().transform(point.get_xydata())``

Cadair replied:
That didn't work either, I can't see any other sensible solution other than this here :confused:

bsipocz commented on astropy/visualization/wcsaxes/core.py:
```diff
@@ -289,12 +290,16 @@ def plot_coord(self, *args, **kwargs):
             if isinstance(frame0, SkyCoord):
                 frame0 = frame0.frame
 
+            native_frame = self._transform_pixel2world.frame_out
```
This doesn't work on 3.2.2:
```
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <astropy.visualization.wcsaxes.core.WCSAxes object at 0x13297d210>
args = (<SkyCoord (ICRS): (ra, dec) in deg
    (359.76045223, 0.26876217)>, 'o')
kwargs = {'transform': <matplotlib.transforms.CompositeGenericTransform object at 0x131f72190>}
frame0 = <ICRS Coordinate: (ra, dec) in deg
    (359.76045223, 0.26876217)>

    def plot_coord(self, *args, **kwargs):
        """
        Plot `~astropy.coordinates.SkyCoord` or
        `~astropy.coordinates.BaseCoordinateFrame` objects onto the axes.
    
        The first argument to
        :meth:`~astropy.visualization.wcsaxes.WCSAxes.plot_coord` should be a
        coordinate, which will then be converted to the first two parameters to
        `matplotlib.axes.Axes.plot`. All other arguments are the same as
        `matplotlib.axes.Axes.plot`. If not specified a ``transform`` keyword
        argument will be created based on the coordinate.
    
        Parameters
        ----------
        coordinate : `~astropy.coordinates.SkyCoord` or `~astropy.coordinates.BaseCoordinateFrame`
            The coordinate object to plot on the axes. This is converted to the
            first two arguments to `matplotlib.axes.Axes.plot`.
    
        See Also
        --------
    
        matplotlib.axes.Axes.plot : This method is called from this function with all arguments passed to it.
    
        """
    
        if isinstance(args[0], (SkyCoord, BaseCoordinateFrame)):
    
            # Extract the frame from the first argument.
            frame0 = args[0]
            if isinstance(frame0, SkyCoord):
                frame0 = frame0.frame
    
>           native_frame = self._transform_pixel2world.frame_out
E           AttributeError: 'WCSAxes' object has no attribute '_transform_pixel2world'

astropy/visualization/wcsaxes/core.py:295: AttributeError
```

bsipocz replied:
(relies on the reorganization from https://github.com/astropy/astropy/pull/8874)
